library/core/src/array/mod.rs

```diff
@@ -21,6 +21,73 @@ mod iter;
 #[stable(feature = "array_value_iter", since = "1.51.0")]
 pub use iter::IntoIter;
 
+/// Creates an array `[T; N]` where each array element `T` is returned by the `cb` call.
+///
+/// # Arguments
+///
+/// * `cb`: Callback where the passed argument is the current array index.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(array_from_fn)]
+///
+/// let array = core::array::from_fn(|i| i);
+/// assert_eq!(array, [0, 1, 2, 3, 4]);
+/// ```
+#[inline]
+#[unstable(feature = "array_from_fn", issue = "none")]
+pub fn from_fn<C, T, const N: usize>(cb: C) -> [T; N]
+where
+    C: FnMut(usize) -> T,
+{
+    // SAFETY: we know for certain that this iterator will yield exactly `N`
+    // items.
+    unsafe { collect_into_array_unchecked(&mut (0..N).map(cb)) }
+}
+
+/// Creates an array `[T; N]` where each fallible array element `T` is returned by the `cb` call.
+/// Unlike `core::array::from_fn`, where the element creation can't fail, this version will return an error
+/// if any element creation was unsuccessful.
+///
+/// # Arguments
+///
+/// * `cb`: Callback where the passed argument is the current array index.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(array_from_fn)]
+///
+/// #[derive(Debug, PartialEq)]
+/// enum SomeError {
+///     Foo,
+/// }
+///
+/// let array = core::array::try_from_fn(|i| Ok::<_, SomeError>(i));
+/// assert_eq!(array, Ok([0, 1, 2, 3, 4]));
+///
+/// let another_array = core::array::try_from_fn::<_, SomeError, (), 2>(|_| Err(SomeError::Foo));
+/// assert_eq!(another_array, Err(SomeError::Foo));
+/// ```
+#[inline]
+#[unstable(feature = "array_from_fn", issue = "none")]
+pub fn try_from_fn<C, E, T, const N: usize>(mut cb: C) -> Result<[T; N], E>
+where
+    C: FnMut(usize) -> Result<T, E>,
+{
+    let mut array: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+    let mut guard: Guard<T, N> =
+        Guard { ptr: MaybeUninit::slice_as_mut_ptr(&mut array), initialized: 0 };
+    for (idx, value) in array.iter_mut().enumerate() {
```

Comment:
Should we be enumerating here? [`iter::from_fn`](https://doc.rust-lang.org/std/iter/fn.from_fn.html) is incredibly similar to `array::from_fn` and doesn't pass the index on each invocation. I'd assume we might want to keep the two consistent? Tracking the index outside the loop would be done in a similar way in both APIs:

```rust
// iter::from_fn
let mut count = 0;
let iter = core::iter::from_fn(|| {
    let out = count;
    count += 1;
    (out < 6).then(|| out)
});
assert_eq!(iter.collect::<Vec<_>>(), [1, 2, 3, 4]);

// array::from_fn
let mut count = 0;
let array = core::array::from_fn(|| {
    let out = count;
    count += 1;
    out
});
assert_eq!(array, [1, 2, 3, 4, 6]);
```

If we _do_ want to keep tracking the index within the closure: we're currently incrementing two counters in the same loop. Rather than calling `enumerate` we could pass the value of `guard.initialized` instead, which already holds the current index.

Reply:
An `iter::from_fn` function is _almost always_ going to be inherently stateful, and iterators don't inherently have indices and a length. On the other hand, an array _does_ have indices and is constructed with a specified finite length — and this means that there is usually some significance to the index of an element, which is frequently useful for computing the element based on its position. Requiring setting up a mutable variable will clutter those common use cases.

If `T::from_fn(F)` functions are to be considered a pattern, then `F`'s parameters should be chosen to reflect the nature of `T`, not just to be similar to each other. Iterators mutate themselves, so the function mutates itself; arrays have indices, so the function will often care about which index it's computing.

Reply:
In addition to what @kpreid said, note that you can also call `enumerate` on iterators while you can't do that with the `from_fn` function for arrays.

Reply:
Hum, I think it is more or less clear that indices are probably useful for these specialized functions but I will raise this concern on the tracking issue.

In regards to keeping track of two different counters, my reasoning is that optimized code is usually good enough and the shared inner function makes it harder to accomplish an unique counter. Nevertheless, such thing can be worked in a following PR.